test/distributed/test_functional_differentials.py

```diff
@@ -0,0 +1,675 @@
+            torch.full((2, 2), fill_value=float(rank), device=device),
+        ]
+        outputs = fcols.all_gather_into_tensor_coalesced(
+            input_tensors, group=group_name
+        )
+
+        # Verify output shapes
+        for output, input_tensor in zip(outputs, input_tensors):
+            expected_shape = list(input_tensor.shape)
+            expected_shape[0] *= self.world_size
+            self.assertEqual(list(output.shape), expected_shape)
+
+    @parametrize("device", devices)
+    def test_reduce_scatter_tensor_coalesced_forward(self, device):
+        """Test reduce_scatter_tensor_coalesced reduces and scatters each tensor.
+
+        Tensors are VARYING (different across ranks).
+        Forward reduces and scatters chunks to ranks.
+        """
+        group_name = dist.group.WORLD.group_name
+        rank = dist.get_rank()
+
+        # Create inputs with appropriate size (divisible by world_size)
+        input_tensors = [
+            torch.full((4 * self.world_size, 3), fill_value=float(rank), device=device),
+            torch.full((2 * self.world_size, 2), fill_value=float(rank), device=device),
+        ]
+        scatter_dims = [0, 0]
+
+        outputs = fcols.reduce_scatter_tensor_coalesced(
+            input_tensors, "sum", scatter_dims, group=group_name
+        )
+
+        # Each rank should receive sum of all ranks' values: 0+1+2+3 = 6
+        expected_value = self.world_size * (self.world_size - 1) / 2
+        for output, input_tensor in zip(outputs, input_tensors):
+            expected_shape = list(input_tensor.shape)
+            expected_shape[0] //= self.world_size
+            self.assertEqual(list(output.shape), expected_shape)
+            expected = torch.full_like(output, fill_value=expected_value)
+            self.assertEqual(output, expected)
+
+    # ============================================================
+    # Phase 2: Backward Correctness Tests
+    # ============================================================
+
+    @parametrize("device", devices)
+    def test_all_reduce_backward(self, device):
+        """Test all_reduce backward does all_reduce.
+
+        Both tensor AND gradients are VARYING (different across ranks).
+        Backward aggregates gradients via all_reduce(sum).
+        """
+        group_name = dist.group.WORLD.group_name
+
+        input_tensor = torch.randn(3, 3, requires_grad=True, device=device)
+        output = fcols.all_reduce(input_tensor, "sum", group=group_name)
+
+        # Backward with ones
+        output.sum().backward()
+
+        # Gradient should be aggregated (backward is all_reduce)
+        expected_grad = torch.full(
+            (3, 3), fill_value=float(self.world_size), device=device
+        )
+        self.assertEqual(input_tensor.grad, expected_grad)
+
+        # Backward is all_reduce (sum)
+        grad_outputs = torch.rand_like(output, device=device)
+        (grad_input,) = torch.autograd.grad(
```

Comment:
how does this work? You shouldn't be able to backward for a second time, after you've done `output.sum().backward()`